### PR TITLE
Mpris blacklist support

### PR DESCRIPTION
**Pull request by abmantis**
Continuation of PR #324.

### CODE REVIEW

**Comment by abmantis** [comment]
When I run the formatted, there are quite a few changes to the `mpris.vala` file. I am assuming the style rule was updated meanwhile. Let me know if you want me to commit those changes too.

**Comment by MrPenguin07** [comment]
> Continuation of PR #324.

Great to see progress on this.... and to see where my code was convoluted :)
I'll pull this now and report back.

Regards

**Comment by MrPenguin07** [comment]
Alright, so with config as so;
```
"mpris": {
    "image-size": 96,
    "image-radius": 6,
    "blacklist": [".*firefox.*"]
},
 ```
 SwayNC reports at startup;
 ```
 ** Message: 13:13:39.774: mpris.vala:263: "org.mpris.MediaPlayer2.firefox.instance_1_5133" is blacklisted
** Message: 13:13:39.774: factory.vala:44: Loading widget: mpris
```

So regex is being respected, and the annoying firefox tabs each creating their own instance is no longer showing in SwayNC.

Thumbsup, I approve.

**Comment by ErikReider** [comment]
I'll get to this after the upcoming bug release :)

**Comment by MrPenguin07** [comment]
I've noticed that firefox instances aren't always blacklisted from the mpris module;

Upon swaync startup, instances are identified and removed. Great.
However, sometimes an existing swaync session does not pickup **new** firefox tabs/mpris players.
When this happens, if I run `$ swaync-client --reload-config` the new firefox tabs are then removed from the mpris player.
I also noticed this happening with my original PR.

I've yet to produce a repeatable scenario for this, however have noticed it happening quite a few times.

**Comment by ErikReider** [comment]
> I've noticed that firefox instances aren't always blacklisted from the mpris module;
> 
> Upon swaync startup, instances are identified and removed. Great. However, sometimes an existing swaync session does not pickup **new** firefox tabs/mpris players. When this happens, if I run `$ swaync-client --reload-config` the new firefox tabs are then removed from the mpris player. I also noticed this happening with my original PR.
> 
> I've yet to produce a repeatable scenario for this, however have noticed it happening quite a few times.

I'll wait until this is resolved then

**Comment by abmantis** [comment]
> > I've noticed that firefox instances aren't always blacklisted from the mpris module;
> > Upon swaync startup, instances are identified and removed. Great. However, sometimes an existing swaync session does not pickup **new** firefox tabs/mpris players. When this happens, if I run `$ swaync-client --reload-config` the new firefox tabs are then removed from the mpris player. I also noticed this happening with my original PR.
> > I've yet to produce a repeatable scenario for this, however have noticed it happening quite a few times.
> 
> I'll wait until this is resolved then

May I suggest that we merge this still? It still improves the functionality since currently, FF will always be duplicated anyway. Also we avoid having stale PRs for things that are already done (even if there is some unknown issue).

If there is some quirk, it can be fixed later. It would also help having more people use this and reporting so we can pinpoint the issue.

Otherwise, we risk never finding the cause and this PR will just stay open forever.

**Comment by MrPenguin07** [comment]
> > I've noticed that firefox instances aren't always blacklisted from the mpris module;
> > Upon swaync startup, instances are identified and removed. Great. However, sometimes an existing swaync session does not pickup **new** firefox tabs/mpris players. When this happens, if I run `$ swaync-client --reload-config` the new firefox tabs are then removed from the mpris player. I also noticed this happening with my original PR.
> > I've yet to produce a repeatable scenario for this, however have noticed it happening quite a few times.
> 
> I'll wait until this is resolved then

I believe, without reviewing the code but from the observed behavior, that this is simply caused by the mpris widget lacking a hook which checks if new players are in the blacklist prior to adding them first.
Thus starting/reloading Swaync will correctly identify and blacklist existing players on startup, however especially with firefox where new instances/players come and go, we lack the mechanism to cross reference the blacklist before adding it.

@abmantis makes some points, though can also accept it's best a feature is complete before merge.
Unless you beat me to it, i'll get onto this later this week regardless.

**Comment by ErikReider** [comment]
I'd rather not merge incomplete features

**Comment by abmantis** [comment]
@MrPenguin07  any findings?

**Comment by abmantis** [comment]
@ErikReider I was now able to reproduce the issue mentioned by @MrPenguin07 and just fixed it.
@MrPenguin07 can you try out my last commit?

**Comment by MrPenguin07** [comment]
> @ErikReider I was now able to reproduce the issue mentioned by @MrPenguin07 and just fixed it. @MrPenguin07 can you try out my last commit?

I appreciate your taking the time to hash this one out, glad to see my suspicion was correct.
Just haven't had time to code much of anything these days :/

I've pulled your latest commit, built swaync and ....
image

Am happy to be greeted by this when a new player is added, without needing to reload swaync.
Good stuff @abmantis 

I consider this feature to now be complete @ErikReider

**Comment by abmantis** [comment]
I initially misinterpreted your description of the issue, that is why I was not able to reproduce it.
Yesterday, after coming back to this and reading it again I noticed that it was in fact easy to reproduce :D 

@ErikReider are you ok with moving this forward now?

**Comment by rtgiskard** [comment]
@abmantis I like this, but I think it's better to keep only necessary changes, `flake.nix` is not likely to be included in this project, and the `message()` can be removed, just to keep it simple

**Comment by rtgiskard** [comment]
Just fixed a bug for `name_owner_changed` signal skipped, this should resolve the issue above, and also this: https://github.com/ErikReider/SwayNotificationCenter/issues/425.

It's now [here](https://github.com/rtgiskard/SwayNotificationCenter) in the main branch with another PR in process, If you're willing, I may create another PR including this feature with the fix.

**Comment by abmantis** [comment]
@rtgiskard the flake.nix was not supposed to have been committed, my bad.
I just noticed that you committed this code directly to `main`. That is fine for this, but IMO doing that is a quick way to push contributors away from wanting to contribute to this project.

**Comment by abmantis** [comment]
Closed by mistake. I thought this was pushed to `main`, but I was checking a fork :facepalm:

File: src/controlCenter/widgets/mpris/mpris.vala [hunk 1]
```diff
@@ -200,6 +213,7 @@ namespace SwayNotificationCenter.Widgets.Mpris {
         }
 
         private void add_player (string name, MprisSource source) {
+            message ("Adding player \"%s\"", name);
```
Comment (abmantis) [inline src/controlCenter/widgets/mpris/mpris.vala:216]:
```suggestion
```

File: src/controlCenter/widgets/mpris/mpris.vala [hunk 2]
```diff
@@ -99,6 +100,16 @@ namespace SwayNotificationCenter.Widgets.Mpris {
                 bool blur_found;
                 bool? blur = get_prop<bool> (config, "blur", out blur_found);
                 if (blur_found) mpris_config.blur = blur;
+
+                // Blacklist
```
Comment (abmantis) [inline src/controlCenter/widgets/mpris/mpris.vala:104]:
```suggestion
```

**Comment by abmantis** [comment]
Removed the flake.nix file, which was committed by mistake.

**Comment by abmantis** [comment]
> Just fixed a typo-like bug for `name_owner_changed` signal handler skipped, which should resolve the issue above, and also this: #425.
> 
> It's now [here](https://github.com/rtgiskard/SwayNotificationCenter) in the main branch with another PR in process, If you're willing, I may create another PR including this feature with the fix.

I think it would be better to keep this PR and to have separate PRs.

**Comment by MrPenguin07** [comment]
Reporting: 2 weeks of testing these commits and the mpris blacklist continues to work perfectly.

File: src/controlCenter/widgets/mpris/mpris.vala [hunk 3]
```diff
@@ -174,6 +184,7 @@ namespace SwayNotificationCenter.Widgets.Mpris {
             string[] names = dbus_iface.list_names ();
             foreach (string name in names) {
                 if (!name.has_prefix (MPRIS_PREFIX)) continue;
+                if (is_blacklisted(name)) continue;
                 if (check_player_exists (name)) return;
```
Comment (rtgiskard) [inline src/controlCenter/widgets/mpris/mpris.vala:188]: You may change this line to fix the unintended return

Comment (andrewsebek) [reply src/controlCenter/widgets/mpris/mpris.vala:188]: Maybe I'm too ignorant about it, I'm seeing that my report was merged in this PR, but I don't know how to use info written here to fix the issue with my Swaync. I uninstalled the one I got from Arch repo, and compiled this one, still having the same issue: Swaync shows MPRIS module when nothing is being played, and when there's something being played, I have two different MPRIS modules in Swaync. What I am missing?

Thanks in advance.

Comment (MrPenguin07) [reply src/controlCenter/widgets/mpris/mpris.vala:188]: > Maybe I'm too ignorant about it, I'm seeing that my report was merged in this PR, but I don't know how to use info written here to fix the issue with my Swaync. I uninstalled the one I got from Arch repo, and compiled this one, still having the same issue: Swaync shows MPRIS module when nothing is being played, and when there's something being played, I have two different MPRIS modules in Swaync. What I am missing?
> 
> Thanks in advance.

By "compiled this one" do you mean you've cloned SwayNotificationCenter repo, pulled the commits to your repo and then built that?
Or are you manually patching the source?

I get the feeling you may simply be building `master` from this repo without the changes.

Comment (andrewsebek) [reply src/controlCenter/widgets/mpris/mpris.vala:188]: Yeah, just cloned and built it. No commits pulled. I don't know how to pull the commits separately, I tried to just copy/paste the changes but apparently I did it wrongly, and I wasn't able to modify the installation...

Sorry for the ignorance :(

**Comment by abmantis** [comment]
@ErikReider  anything else required for this?

**Comment by andrewsebek** [comment]
> @ErikReider  anything else required for this?

Should be merged, it works flawlessly.

File: man/swaync.5.scd [hunk 4]
```diff
@@ -595,7 +604,8 @@ config file to be able to detect config errors
 		},
 		"mpris": {
 			"image-size": 96,
-			"image-radius": 12
+			"image-radius": 12,
+			"blacklist": ["playerctld"]
```
Comment (ErikReider) [inline man/swaync.5.scd:608]: Wouldn't this be `org.mpris.MediaPlayer2.playerctld`? That's at least what appears on my machine

File: src/controlCenter/widgets/mpris/mpris.vala [hunk 5]
```diff
@@ -99,6 +100,15 @@ namespace SwayNotificationCenter.Widgets.Mpris {
                 bool blur_found;
                 bool? blur = get_prop<bool> (config, "blur", out blur_found);
                 if (blur_found) mpris_config.blur = blur;
+
+                Json.Array ? blacklist = get_prop_array (config, "blacklist");
+                if (blacklist != null) {
+                    string[] apps = new string[blacklist.get_length ()];
+                    for (int i = 0; i < blacklist.get_length (); i++) {
+                        apps[i] = blacklist.get_string_element (i);
```
Comment (ErikReider) [inline src/controlCenter/widgets/mpris/mpris.vala:108]: Nit: Should double-check that it's a string

File: src/controlCenter/widgets/mpris/mpris.vala [hunk 6]
```diff
@@ -99,6 +100,15 @@ namespace SwayNotificationCenter.Widgets.Mpris {
                 bool blur_found;
                 bool? blur = get_prop<bool> (config, "blur", out blur_found);
                 if (blur_found) mpris_config.blur = blur;
+
+                Json.Array ? blacklist = get_prop_array (config, "blacklist");
+                if (blacklist != null) {
+                    string[] apps = new string[blacklist.get_length ()];
```
Comment (ErikReider) [inline src/controlCenter/widgets/mpris/mpris.vala:106]: Why not initialize the `mpris_config.blacklist` array instead?

File: man/swaync.5.scd (hunk 4 above)
Comment (ErikReider) [reply man/swaync.5.scd:608]: Or is this due to the blacklist checking with regex?

**Review by ErikReider — CHANGES_REQUESTED** [review]
Other than that, LGTM! Thanks for your extreme patience xD

File: man/swaync.5.scd (hunk 4 above)
Comment (abmantis) [reply man/swaync.5.scd:608]: yeah, because regex

File: src/controlCenter/widgets/mpris/mpris.vala (hunk 6 above)
Comment (abmantis) [reply src/controlCenter/widgets/mpris/mpris.vala:106]: done

File: src/controlCenter/widgets/mpris/mpris.vala (hunk 5 above)
Comment (abmantis) [reply src/controlCenter/widgets/mpris/mpris.vala:108]: done

**Comment by MrPenguin07** [comment]
LGTM :)

**Review by ErikReider — APPROVED** [review]
LGTM! Thanks! :D

**Comment by andrewsebek** [comment]
Sorry for bumping this again, but I have to comment something:

I reinstalled Arch Linux and installed my dotfiles one more time a couple days ago, and installed Swaync from the main repo (not building it from source) and the problem persists.

My code in mpris:

"mpris": {
      "image-size": 60,
      "image-radius": 12,
      "blacklist":["playerctld"]
    },

Am I missing something? I saw it was merged already, so I was expecting this correction to be available in Arch repos already.

**Comment by MrPenguin07** [comment]
> Sorry for bumping this again, but I have to comment something:

> Am I missing something? I saw it was merged already, so I was expecting this correction to be available in Arch repos already.

```
"mpris": {
"image-size": 60,
"image-radius": 12,
"blacklist":["playerctld"]
},
```

Presuming you've copy-pasted, try adding a space between `"blacklist":` and `["playerctld"]`

**Comment by ErikReider** [comment]
> Sorry for bumping this again, but I have to comment something:
> 
> I reinstalled Arch Linux and installed my dotfiles one more time a couple days ago, and installed Swaync from the main repo (not building it from source) and the problem persists.
> 
> My code in mpris:
> 
> "mpris": { "image-size": 60, "image-radius": 12, "blacklist":["playerctld"] },
> 
> Am I missing something? I saw it was merged already, so I was expecting this correction to be available in Arch repos already.

It's not in a current release yet, you'll need to install swaync-git from the AUR

**Comment by andrewsebek** [comment]
> > Sorry for bumping this again, but I have to comment something:
> > 
> > I reinstalled Arch Linux and installed my dotfiles one more time a couple days ago, and installed Swaync from the main repo (not building it from source) and the problem persists.
> > 
> > My code in mpris:
> > 
> > "mpris": { "image-size": 60, "image-radius": 12, "blacklist":["playerctld"] },
> > 
> > Am I missing something? I saw it was merged already, so I was expecting this correction to be available in Arch repos already.
> 
> It's not in a current release yet, you'll need to install swaync-git from the AUR

It totally worked on the -git version. I'll be there until the stable branch is ready. Thanks for the response!

**Pull request merged**